<a href="https://colab.research.google.com/github/sonjoy1s/ML/blob/main/Assignment_01_Machine_Learning_Course.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Week 01 Assignment  
## Data Quality, Evaluation, Scaling, and Encoding

**Student name: ** Sonjoy Chandra

This is a small assignment that connects topics from Module 1, 2, and 3.  
You must complete it in this Colab notebook.

You will need to use concepts that appeared in the videos:
- Module 1 and 2: basic descriptive statistics, proportions, confusion matrix, accuracy, precision, recall
- Module 3: standardization, min max scaling, nominal vs ordinal, one hot encoding, ordinal encoding, Euclidean and Manhattan distance

Please do not use any extra libraries beyond `pandas`, `numpy`.



---
## 0. Setup and Dataset

We will use a dataset that should have columns given below:

- `user_id`  
- `age`  
- `monthly_income` (numeric)  
- `daily_screen_time_min` (numeric)  
- `daily_app_opens` (numeric)  
- `true_label` and `pred_label` for a binary classification task (0 or 1)  
- `satisfaction_level` (for example: `Low`, `Medium`, `High`)  
- `city_type` (for example: `Urban`, `Suburban`, `Rural`)


In [153]:
# Cell 1: Imports
import pandas as pd
import numpy as np

In [154]:
# Cell 2: Load the dataset (Already done for you)
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220")

# Show first few rows
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban


### 0.1 Check your dataset

1. Confirm that the dataset loaded correctly.  
2. Check that you have at least these columns:  
   - numeric: `age`, `monthly_income`, `daily_screen_time_min`, `daily_app_opens`  
   - labels: `true_label`, `pred_label`  
   - categorical: `satisfaction_level`, `city_type`  



---
## Part A - Module 1 and 2 Review

In this part you will do simple descriptive statistics and basic classification evaluation.



### Q1. Descriptive statistics on a numeric feature

Choose one numeric column, for example `daily_screen_time_min`.


In [155]:
# Q1.1: Choose your numeric column here [We already write this ans]
num_col = "daily_screen_time_min"

df[num_col].describe()

count    100.000000
mean     181.890000
std       68.886951
min       60.000000
25%      122.000000
50%      178.000000
75%      243.750000
max      299.000000
Name: daily_screen_time_min, dtype: float64


> **Q1.2 Short answer: [Marks: 05]**  
> Look at the count, mean, min, max, and standard deviation for your chosen column.  
> In 2 to 3 sentences, comment on what you see.  
> For example, does the max look very far from the mean, or does it look quite close?

Write your answer here:

>  Count : Find the how many rows.Dataset rows counts 100.
> Mean : Find the average value. Dataset mean value 181.890000
>  Min : Find the lowest value. Dataset min value 60
> Max : Find the highest value.Dataset max value 299
>Standard deviation : How spread out is the data.



### Q2. Proportion of positive class

Use the `true_label` column, where 1 means "positive" and 0 means "negative".


In [156]:
# Q2.1: Compute proportion of positive class [We already write this ans]
label_col = "true_label"

positive_count = (df[label_col] == 1).sum()
total_count = df.shape[0]
positive_proportion = positive_count / total_count

print("Positive count:", positive_count)
print("Total samples:", total_count)
print("Proportion of positive class:", positive_proportion)

Positive count: 52
Total samples: 100
Proportion of positive class: 0.52



> **Q2.2 Short answer: [5 marks]**  
> In 1 to 2 sentences, explain what this proportion tells you about your dataset.  
> For example, is the dataset balanced between 0 and 1, or is one class much more common?

Write your answer here:
> Ans :
>Code Explain: Frist line find, how many rows positive . Second line how many total samples or rows in dataset.3rd line find the positive average value.
> This proportion find how many rows in the dataset positive class .ex: true_label = 1.
If the proportion is low it means the 1 class is very small and the 0 class is much larger the dataset a imbalanced with the 1 class much less common than the 0 class. In other words the dataset is not balanced between 0 and 1.




### Q3. Confusion matrix and basic metrics

For this question, use:
- `true_label` as the actual label  
- `pred_label` as the model prediction


In [157]:
# Q3.1: Manually compute TP, TN, FP, FN [We already write this ans]
true_col = "true_label"
pred_col = "pred_label"

tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 28
TN: 27
FP: 21
FN: 24


In [158]:
# Q3.2: Compute accuracy, precision, recall [We already write this ans]
accuracy = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0

precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.55
Precision: 0.5714285714285714
Recall: 0.5384615384615384



> **Q3.3 Short answer: [10 marks]**  
> In 3 to 4 sentences, briefly comment on the model using these three metrics.  
> For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

Write your answer here:

>Ans : Accuracy 0.55, precision 0.572 if positivly predict 57% correct and recall 0.54 actual detect 54% positive that means model half positive value miss and some false positive




---
## Part B - Module 3: Scaling and Encoding

Now we will pick a few features and apply scaling and encoding.



### Q4. Standardization and Min max scaling

Use one numeric column, `monthly_income`.


In [159]:
# Q4.1: Choose the numeric column [2 marks]
# Ans :
num_col = "monthly_income"
#df.head(2)

In [160]:
# Q4.2: Standardization with z-score [10 marks]
# Ans :
# formula: z = (x-mean)/sd
mean = df[num_col].mean()
std = df[num_col].std()
z = (df[num_col] - mean) / std
df['monthly_income'] = z.round(2)
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43            0.94                    109               48   
1        2   49           -0.32                    194                7   
2        3   19            0.74                    146               36   
3        4   19            1.04                    287               14   
4        5   63           -1.26                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban

In [161]:
# Q4.3: Min max scaling implementation [10 marks]
# Ans :
# Formula : x = (x-xmin)/(xmax - xmin)
min = df[num_col].min()
max = df[num_col].max()
x = (df[num_col] - min) / (max - min)
df['monthly_income'] = x.round(2)
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43            0.67                    109               48   
1        2   49            0.39                    194                7   
2        3   19            0.63                    146               36   
3        4   19            0.70                    287               14   
4        5   63            0.19                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban


> **Q4.4 Short answer: [3 marks]**  
> Compare the standardized and min max scaled columns in 2 to 3 sentences.  
> Mention what kind of range each one uses and how the numbers look.

Write your answer here:

>  standardized/ z score : Standardization centers data at 0 with a std of 1, so values can be negative or positive.
#is not ans just note : standardize data ke 0 mean ar 1 std te niye ase tai value neg o pos hote pare
>  Min-Max scaled : Min-max scaling rescales data to a fixed range(0-1), keeping all values positive within the bounds.
#is not a ans just note : min-max scaling data ke 0-1 range e niye ase sob ghula value positive ar defined bounds er moddhe thake



### Q5. One hot and ordinal encoding

We will use:
- `city_type` as a nominal feature  
- `satisfaction_level` as an ordinal feature with order `Low` < `Medium` < `High`  


In [162]:
# Q5.1: One hot encoding for city_type using pandas [10 marks]
# Ans :
col_city = pd.get_dummies(df, columns=['city_type'],prefix='new',dtype=int)
#col_city
col_city.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43            0.67                    109               48   
1        2   49            0.39                    194                7   
2        3   19            0.63                    146               36   
3        4   19            0.70                    287               14   
4        5   63            0.19                     66               46   

   true_label  pred_label satisfaction_level  new_Rural  new_Suburban  \
0           0           0             Medium          0             1   
1           0           0                Low          0             0   
2           1           0               High          1             0   
3           1           0               High          0             1   
4           0           0             Medium          0             1   

   new_Urban  
0          0  
1          1  
2          0  
3          0  
4          0

In [163]:
# Q5.2: Attach one hot encoded columns to df [5 marks]
# Ans :
df_encoding = pd.concat([df,col_city],axis=1)
df_encoding=df_encoding.drop(['city_type'],axis=1)
#df_encoding
df_encoding.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43            0.67                    109               48   
1        2   49            0.39                    194                7   
2        3   19            0.63                    146               36   
3        4   19            0.70                    287               14   
4        5   63            0.19                     66               46   

   true_label  pred_label satisfaction_level  user_id  age  monthly_income  \
0           0           0             Medium        1   43            0.67   
1           0           0                Low        2   49            0.39   
2           1           0               High        3   19            0.63   
3           1           0               High        4   19            0.70   
4           0           0             Medium        5   63            0.19   

   daily_screen_time_min  daily_app_opens  true_label  pred_label  \
0                    109               48           0           0   
1                    194                7           0           0   
2                    146               36           1           0   
3                    287               14           1           0   
4                     66               46           0           0   

  satisfaction_level  new_Rural  new_Suburban  new_Urban  
0             Medium          0             1          0  
1                Low          0             0          1  
2               High          1             0          0  
3               High          0             1          0  
4             Medium          0             1          0

In [164]:
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]
# Ans :
order= {'Low': 0, 'Medium': 1, 'High': 2}
df['satisfaction_level'] = df['satisfaction_level'].map(order).astype(int)
#df['satisfaction_level']=df['satisfaction_level_encoded']
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43            0.67                    109               48   
1        2   49            0.39                    194                7   
2        3   19            0.63                    146               36   
3        4   19            0.70                    287               14   
4        5   63            0.19                     66               46   

   true_label  pred_label  satisfaction_level city_type  
0           0           0                   1  Suburban  
1           0           0                   0     Urban  
2           1           0                   2     Rural  
3           1           0                   2  Suburban  
4           0           0                   1  Suburban


> **Q5.4 Short answer: [5 marks]**  
> In 2 to 3 sentences, explain why one hot encoding is suitable for `city_type`  
> and why ordinal encoding is suitable for `satisfaction_level`.

Write your answer here:

>Ans :
>  City_type : One-hot encoding suitable for city_type nominal avoid false order.
> satisfaction_level : ordinal encoding right satisfaction_lavel ordinal , it's natural order(Low < Medium < High) .



---
## Part C - Module 3: Distances between users

For this small part we will work with vectors based on scaled numeric features.



### Q6. Euclidean and Manhattan distance

Build 2D vectors for user 0 and user 1 using:
- `income_std`  
- `daily_app_opens` (or its min max scaled version if you prefer)


In [165]:
# Q6.1: Build 2D vectors for first two users [We already write this ans]
vec_cols = ["monthly_income", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values
v2 = df.loc[1, vec_cols].values

print("v1:", v1)
print("v2:", v2)

v1: [np.float64(0.67) np.int64(48)]
v2: [np.float64(0.39) np.int64(7)]


In [166]:
# Q6.2: Euclidean distance computation [10 marks]
# Ans :
# Formula : d2(p,q)=rut(p-q)**2
dis = np.linalg.norm(v1-v2)
print(dis.round(3))

41.001


In [167]:
# Q6.3: Manhattan distance computation [10 marks]
# Ans :
# Formula : d1(p,q)=|p1-q1|+|p2-q2|
manh = np.linalg.norm(v1-v2,ord=1)
print(manh)

41.28



> **Q6.4 Short answer: [5 marks]**  
> Which one is larger in your result, Euclidean or Manhattan distance  
> and why does that usually happen based on their formulas?

Write your answer here:

>  Ans : Manhattan > Euclidean because Euclidean measures the straight line distance. Manhattan abs(+) difference along each dimension.


> #not a ans just note: Manhattan Euclidean ceye boro.  
>  



---
## Final Reflection [5 marks]

> In 4 to 6 sentences, describe how the three modules connect in this assignment.  
> Mention:
> - One idea from Module 1 or 2 that you used  
> - One idea from Module 3 that you used  
> - How these ideas together help you understand a dataset more deeply

Write your reflection here:

>  ans :
>  module 1-2 : Accuracy,precision,recall for model performance.
>  module 3 : Apply one-hot and ordinal encoding for categorical data to numerical format.
> This process help for analysis or machine learning.



## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Download this notebook as `.ipynb` and upload it according to the given instructions.
- ***Must Read Assignment Module Text Instruction fully Where you will find how to submit this assignment***
